### PR TITLE
En passant (includes move_to method & db update)

### PR DESCRIPTION
**Pull request by kellylougheed**
- Updated the pieces model so that the default value of pieces.moves = 0.

- Wrote move_to method that implements capture_piece and a new method, en_passant. The move_to method updates the piece's x_position, y_position, and moves (until two moves have been made and the count is no longer of concern).

- Wrote an en_passant method that allows a piece to capture a pawn that moved two spaces forward in its first move by moving to the space that the pawn skipped over.

- Wrote two tests each for move_to and en_passant which are all passing.

### CODE REVIEW

**Comment by PamBWillenz** [comment]
@kellylougheed - looks good. I see the `move_to` includes a capture portion - specifically for the En Passant for the pawn- will I be able to use this for the king? The castling doesn't have a capture portion so wondering if the method will work for just the move part or I will need to create another method in piece.rb for just 'move_to`? Or you can create or put the capture portion of the`move_to` into the En Passant method?

**Comment by PamBWillenz** [comment]
sounds good @kellylougheed. I'll have to define `castle` in piece.rb to do that, right? I can't borrow from the king.rb to the piece.rb, or can I? You can only inherit from the super class not from a child class (maybe you can). Let me  know what you think?

**Comment by kellylougheed** [comment]
@PamBWillenz Yeah, I think castling will have to be in piece.rb so it involves the king & the rook. I think you can only inherit from a super class.

If you think the code looks good, feel free to approve it! :) or I'm happy to answer any questions about it.

**Comment by PamBWillenz** [comment]
I will approve @kellylougheed - another question about castling. I can define castling (and there are two other methods I think I need and other pieces may need too so I will put them in piece.rb) in piece.rb and then use the two methods `castle!` and `can_castle?` in king.rb. This is my strategy, so far. Thoughts?

**Comment by kellylougheed** [comment]
That sounds good @PamBWillenz! Breaking it up into other methods (which you can call on the king & rook) sounds good.

**Comment by PamBWillenz** [comment]
Another question, @kellylougheed - I want to understand the move_counter (which is different from our datatype `moves`), you set it here for 1 only if that piece (the pawn) can move only one. The move counter can be set to how the piece moves, right? The king can only move one too, but in castling, the rook and king move more than 1. Checking the versatility of this variable.
```  move_counter = self.moves
     move_counter += 1 if self.moves < 2```

Thanks for answering these questions, Kelly. Your code is amazing and I want to try and comprehend it as best as I can. Learning a lot from reading your code :smile_cat:

**Comment by kellylougheed** [comment]
Thanks, @PamBWillenz! And no worries! The move_counter is just a local variable I created in case "moves" needs to updated in the database (only necessary if the piece has moved less than twice -- we're only using this for en_passant and castling, I think). If the piece has moved less than two times, I want to be sure to increment its "moves" count in the database, so I store that increment in move_counter before I pass it to the database. Otherwise, move_counter is just the value of "moves" and it doesn't update the database. Actually, I suppose I could make a conditional that won't even send anything to the database if "moves" is already at 2! Hold on...

**Comment by kellylougheed** [comment]
OK, code updated (plus another test)! It shouldn't send anything regarding the move count to the database if the piece has moved at least twice, because 0 moves / 1 move / 2+ moves are really all we care about.

**Comment by kellylougheed** [comment]
Thanks @PamBWillenz!

**Comment by PamBWillenz** [comment]
Hey Kelly @kellylougheed - I don't think I'll be able to put any castling logic in the `move_to` method in piece.rb. I can certainly access it and use like the `valid_move` like we did for each piece. The castling king has more conditionals than the pawn so I have to reference several methods (6 or 7) just to define `can_castle` and `castle`. I don't think we want this much code in the piece.rb. I think it belongs in the king.rb. I'll wait till your PR gets merged before I push my castling PR since I will definitely use `move_to`. If you have any ideas how else to do it, please let me know.

**Comment by kellylougheed** [comment]
OK @PamBWillenz! Looking forward to seeing how the castling works :) Definitely feel free to mess around with and modify move_to as well!

File: app/models/piece.rb [hunk 1]
```diff
@@ -18,6 +18,38 @@ def capture_piece(x, y)
...
+  end
+
+  def en_passant(destination_x, destination_y)
+    if destination_y == 2 || destination_y == 5
+      destination_y == 2 ? incrementer = 1 : incrementer = -1
+      incremented_y = destination_y + incrementer
+      if self.game.is_piece_present?(destination_x, incremented_y)
+        other_piece = game.get_piece(destination_x, incremented_y)
+        if other_piece.type == 'Pawn' && 
+          other_piece.moves == 1 &&
```
Comment (FsDevNinja) [inline app/models/piece.rb]: Is this moves column tracking how many moves a specific piece has made? Or how many moves all pieces have made?

Comment (kellylougheed) [inline app/models/piece.rb]: Each piece has a moved column that counts how many moves that piece has made (up to 2 because that is all we care about).

Comment (kellylougheed) [inline app/models/piece.rb]: *moves column, not "moved" column. Phone...

**Comment by PamBWillenz** [comment]
Hi @kellylougheed Kelly -  I know I've asked this but I am still wondering about the 2 move limit - is that true for the first move - for either pawn/king - or for any move after a first move? For castling, both the king and the rook have to be using their first move to castle. I think `valid_move` then takes over so the rook can move its four spaces to castle - correct?

**Comment by kellylougheed** [comment]
@PamBWillenz not sure I understand what you're asking but a piece will have zero moves before it makes the first move, one move after it makes its first move, and two moves after it has moved twice. So for castling, you would have to check that they both have zero moves. I think you'll have to override valid move because it will return invalid if there is a piece of the same color at the destination square.

**Comment by PamBWillenz** [comment]
Ok - so I would override `valid_move` vs `is_obstructed`? The move is different because they are almost moving simultaneously.

**Comment by kellylougheed** [comment]
@PamBWillenz Maybe valid_move since it calls is_obstructed!

**Pull request merged**